### PR TITLE
backend/feat/#993 : Made connection of  booking , search request and ride table with location and location mapping table

### PR DESCRIPTION
**Pull request by JanhaviSoni**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

### Additional Changes

- [x] This PR modifies the database schema (database migration added)
- [x] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Checklist

- [ ] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable
- [ ] No leak detected in leakcanary

### CODE REVIEW

File: Backend/dev/migrations/rider-app/1055-merchant-maps-configs.sql [hunk 1]
```diff
@@ -29,9 +29,9 @@ INSERT INTO atlas_app.merchant_service_usage_config (merchant_id, get_distances,
 -- googleKey = "mock-google-key"
 WITH MerchantMapsServiceConfigs AS (
   SELECT T1.id, 'Maps_Google', CAST ('{
-   "googleMapsUrl":"http://localhost:8019/",
-   "googleRoadsUrl":"http://localhost:8019/",
-   "googleKey":"0.1.0|0|Cb8TKe1UWD9IkUDfVg4eE+moqGGHZIlhvXmWDqG/kS1RgdpRQF5Wxigi762BJhW5dKS9iHNnFKrZ2dm+vpuT61Bp"
+   "googleMapsUrl":"https://maps.googleapis.com/maps/api/",
+   "googleRoadsUrl":"https://roads.googleapis.com/",
+   "googleKey":"0.1.0|2|S34+Lq69uC/hNeYSXr4YSjwwmaTS0jO/1ZGlAAwl72hBhgD9AAZhgI4o/6x3oi99KyJkQdt5UvgjlHyeEOuf1Z3xzOBqWBYVQM/RBggZ7NggTyIsDgiG5b3p"
   }' AS json)
```
Comment (Anuragini-2506) [inline Backend/dev/migrations/rider-app/1055-merchant-maps-configs.sql:35]: These are pushed by mistake ?

File: Backend/dev/migrations/dynamic-offer-driver-app/0051-merchant-maps-configs.sql [hunk 2]
```diff
@@ -29,9 +29,9 @@ INSERT INTO atlas_driver_offer_bpp.merchant_service_usage_config (merchant_id, g
 -- googleKey = "mock-google-key"
 WITH MerchantMapsServiceConfigs AS (
   SELECT T1.id, 'Maps_Google', CAST ('{
-   "googleMapsUrl":"http://localhost:8019/",
-   "googleRoadsUrl":"http://localhost:8019/",
-   "googleKey":"0.1.0|0|Cb8TKe1UWD9IkUDfVg4eE+moqGGHZIlhvXmWDqG/kS1RgdpRQF5Wxigi762BJhW5dKS9iHNnFKrZ2dm+vpuT61Bp"
+   "googleMapsUrl":"https://maps.googleapis.com/maps/api/",
+   "googleRoadsUrl":"https://roads.googleapis.com/",
+   "googleKey":"0.1.0|2|S34+Lq69uC/hNeYSXr4YSjwwmaTS0jO/1ZGlAAwl72hBhgD9AAZhgI4o/6x3oi99KyJkQdt5UvgjlHyeEOuf1Z3xzOBqWBYVQM/RBggZ7NggTyIsDgiG5b3p"
```
Comment (Anuragini-2506) [inline Backend/dev/migrations/dynamic-offer-driver-app/0051-merchant-maps-configs.sql:34]: These are pushed by mistake ?
Please revert to original if it is.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/SearchRequestForDriver.hs [hunk 3]
```diff
@@ -92,8 +92,8 @@ data SearchRequestForDriverAPIEntity = SearchRequestForDriverAPIEntity
     durationToPickup :: Seconds,
     baseFare :: Money,
     customerExtraFee :: Maybe Money,
-    fromLocation :: DLoc.SearchReqLocation,
-    toLocation :: DLoc.SearchReqLocation,
+    fromLocation :: DLoc.Location,
+    toLocation :: DLoc.Location, -- we need to show all requests or last one ?
```
Comment (roman-bodavskiy) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/SearchRequestForDriver.hs:96]: It wouldn't backward compatible if we will show all locations. Maybe add separate field for all `toLocations`? @hkmangla

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Beam/Location.hs [hunk 4]
```diff
@@ -12,8 +12,9 @@
   the GNU Affero General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.
 -}
 {-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE TemplateHaskell #-}
```
Comment (roman-bodavskiy) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Beam/Location.hs:15]: `TemplateHaskell` is already there in `package.yaml`

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Beam/LocationMapping.hs [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+ is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+
+ or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more details. You should have received a copy of
+
+ the GNU Affero General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.
+-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TemplateHaskell #-}
```
Comment (roman-bodavskiy) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Beam/LocationMapping.hs:16]: `StandaloneDeriving` and `TemplateHaskell` are in `package.yaml`

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/SearchRequestForDriver.hs (hunk 3 above)
Comment (JanhaviSoni) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/SearchRequestForDriver.hs:96]: Yes @roman-bodavskiy we need to discuss these changes for multiple locations . I have just commented out where I was having doubt since I am expecting to do these changes at a later stage when I will work on adding multiple stops after completing edit pickup and edit destination.

@roman-bodavskiy @hemantmangla please suggest how can we handle them.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/Booking.hs [hunk 6]
```diff
@@ -107,9 +117,16 @@ findFareForCancelledBookings :: MonadFlow m => [Id Booking] -> m Money
...
 instance FromTType' BeamB.Booking Booking where
+  fromTType' :: MonadFlow m => BeamB.Booking -> m (Maybe Booking)
   fromTType' BeamB.BookingT {..} = do
-    fl <- QBBL.findById (Id fromLocationId) >>= fromMaybeM (InternalError $ "FromLocation not found in booking for fromLocationId: " <> show fromLocationId)
-    tl <- QBBL.findById (Id toLocationId) >>= fromMaybeM (InternalError $ "ToLocation not found in booking for toLocationId: " <> show toLocationId)
+    mappings <- QLM.findByEntityId id
+    let fromLocationMapping = filter (\loc -> loc.order == 0) mappings
+        toLocationMappings = filter (\loc -> loc.order /= 0) mappings
+    when (null fromLocationMapping || null toLocationMappings) $ throwError (InternalError "Entity Mappings Not Found")
+    let fromLoc = head fromLocationMapping
```
Comment (roman-bodavskiy) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/Booking.hs:126]: you can avoid partial `head` function:

`listToMaybe fromLocationMapping & fromMaybeM  (InternalError "Entity to location Mappings Not Found")`

or even 
```
case fromLocationMapping of
  [] -> throwError (InternalError "Entity from location mappings not found")
  [x] -> pure x
  _ ->  throwError (InternalError "Multiple from location Mappings)
```

Also proper error codes can be added

File: Backend/dev/migrations/dynamic-offer-driver-app/0260-adding-location-and-location-mapping-tables.sql [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+   atlas_driver_offer_bpp.search_request DROP COLUMN from_location_id;
+
+ALTER TABLE
+   atlas_driver_offer_bpp.search_request DROP COLUMN to_location_id;
+
+ALTER TABLE
+   atlas_driver_offer_bpp.booking DROP COLUMN from_location_id;
+
+ALTER TABLE
+   atlas_driver_offer_bpp.booking DROP COLUMN to_location_id;
```
Comment (roman-bodavskiy) [inline Backend/dev/migrations/dynamic-offer-driver-app/0260-adding-location-and-location-mapping-tables.sql:43]: I guess we should create `location_mapping` entries for old data before removing this columns

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/Booking.hs [hunk 8]
```diff
@@ -107,9 +117,16 @@ findFareForCancelledBookings :: MonadFlow m => [Id Booking] -> m Money
...
+    fl <- QL.findById fromLoc.locationId >>= fromMaybeM (InternalError $ "FromLocation not found in booking for fromLocationId: " <> fromLoc.locationId.getId)
+    let toLocMap = last toLocationMappings
```
[the remark this answers is not in the document]
Comment (JanhaviSoni) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/Booking.hs:128]: Since in future we will require list of toLocationMappings so I guess we can go with last or mb we can sort on the basis of order in findById so that every time mappings are return in proper order

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Beam/LocationMapping.hs [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+module Storage.Beam.LocationMapping where
+
+import Data.Serialize
+import qualified Database.Beam as B
+import Database.Beam.MySQL ()
+import qualified Domain.Types.LocationMapping as DLM
+import EulerHS.KVConnector.Types (KVConnector (..), MeshMeta (..), primaryKey, secondaryKeys, tableName)
+import GHC.Generics (Generic)
+import Kernel.Prelude hiding (Generic)
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Beam/LocationMapping.hs:24]: Same as above file.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/Booking.hs [hunk 10]
```diff
@@ -16,26 +17,39 @@
...
+import qualified Storage.Queries.Location as QL
+import qualified Storage.Queries.LocationMapping as QLM
 
 createBooking :: MonadFlow m => Booking -> m ()
 createBooking = createWithKV
 
 create :: MonadFlow m => Booking -> m ()
-create dBooking = QBBL.create dBooking.fromLocation >> QBBL.create dBooking.toLocation >> createBooking dBooking
+create dBooking = do
+  _ <- whenNothingM_ (QL.findById dBooking.fromLocation.id) $ do QL.create dBooking.fromLocation
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/Booking.hs:47]: Why find first? It will not be there, right?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/Booking.hs [hunk 11]
```diff
@@ -16,26 +17,39 @@
...
+  _ <- whenNothingM_ (QL.findById dBooking.toLocation.id) $ do QL.create dBooking.toLocation
+  createBooking dBooking
+
+createBookingAndMapping :: MonadFlow m => Booking -> DLM.LocationMapping -> DLM.LocationMapping -> m ()
+createBookingAndMapping booking fromLocationMap toLocationMaps = create booking >> QLM.create fromLocationMap >> QLM.create toLocationMaps
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/Booking.hs:52]: first create locationMap then only create booking.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/Location.hs [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+        BeamL.fullAddress = (fullAddress :: LocationAddress -> Maybe Text) address,
+        BeamL.createdAt = createdAt,
+        BeamL.updatedAt = updatedAt
+      }
+
+getBookingLocs ::
+  (MonadFlow m) =>
+  [Id Location] ->
+  m [Location]
+getBookingLocs locationIds = findAllWithKV [Se.Is BeamL.id $ Se.In $ getId <$> locationIds]
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/Location.hs:85]: Define functions above fromTType and toTType conversion.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Beam/SearchRequestSpecialZone.hs [hunk 13]
```diff
@@ -44,8 +44,6 @@ data SearchRequestSpecialZoneT f = SearchRequestSpecialZoneT
     startTime :: B.C f Time.UTCTime,
     validTill :: B.C f Time.UTCTime,
     providerId :: B.C f Text,
-    fromLocationId :: B.C f Text,
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Beam/SearchRequestSpecialZone.hs:47]: Why did you remove for this?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/SearchRequestSpecialZone.hs [hunk 14]
```diff
@@ -54,8 +55,14 @@ getValidTill (Id searchRequestId) = do
 
 instance FromTType' BeamSRSZ.SearchRequestSpecialZone SearchRequestSpecialZone where
   fromTType' BeamSRSZ.SearchRequestSpecialZoneT {..} = do
-    fl <- QSRL.findById (Id fromLocationId) >>= fromMaybeM (InternalError $ "FromLocation not found in SearchRequestSpecialZone for fromLocationId: " <> show fromLocationId)
-    tl <- QSRL.findById (Id toLocationId) >>= fromMaybeM (InternalError $ "ToLocation not found in SearchRequestSpecialZone for toLocationId: " <> show toLocationId)
+    mappings <- QLM.findByEntityId id
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/SearchRequestSpecialZone.hs:58]: Why backward compatibility not handled for this?

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Search/Rental.hs [hunk 15]
```diff
@@ -79,7 +82,11 @@ rentalSearch personId bundleVersion clientVersion device req = do
   Metrics.incrementSearchRequestCount merchant.name
   let txnId = getId (searchRequest.id)
   Metrics.startSearchMetrics merchant.name txnId
+  fromLocationMap <- SLM.buildPickUpLocationMapping searchRequest.fromLocation.id txnId DLM.SEARCH_REQUEST
+  mbToLocationMap <- maybe (pure Nothing) (\detail -> Just <$> SLM.buildDropLocationMapping detail.id txnId DLM.SEARCH_REQUEST) searchRequest.toLocation
```
Comment (hkmangla) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Search/Rental.hs:86]: Same as BPP side comment. Make a common function for it.

File: Backend/dev/migrations/dynamic-offer-driver-app/0260-adding-location-and-location-mapping-tables.sql [hunk 16]
```diff
@@ -0,0 +1,51 @@
+CREATE TABLE atlas_driver_offer_bpp.location (
+   id CHARACTER(36) PRIMARY KEY NOT NULL,
+   lat DOUBLE PRECISION NOT NULL,
+   lon DOUBLE PRECISION NOT NULL,
+   street CHARACTER VARYING(255),
+   door CHARACTER VARYING(255),
+   city CHARACTER VARYING(255),
+   state CHARACTER VARYING(255),
+   country CHARACTER VARYING(255),
+   building CHARACTER VARYING(255),
```
Comment (hkmangla) [inline Backend/dev/migrations/dynamic-offer-driver-app/0260-adding-location-and-location-mapping-tables.sql:10]: Let make all address component as `text` instead of `CHARACTER VARYING(255)`

**Review by hkmangla — CHANGES_REQUESTED** [review]

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/LocationMapping.hs [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+  id <- generateGUID
+  let order = 0
+  return DLM.LocationMapping {..}
+
+buildDropLocationMapping :: MonadFlow m => Id DL.Location -> Text -> DLM.LocationMappingTags -> m DLM.LocationMapping
+buildDropLocationMapping locationId entityId tag = do
+  id <- generateGUID
+  version <- getMappingVersion
+  noOfEntries <- QLM.countOrders entityId
+  let order = if noOfEntries == 0 then 1 else noOfEntries
```
[the remark this answers is not in the document]
Comment (JanhaviSoni) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/LocationMapping.hs:37]: since we are creating fromLocationMapping and toLocationMapping together while createbooking or creatRide then both won't exist so here we need to assign 1 to it . This case is handled here.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/Booking.hs (hunk 10 above)
Comment (JanhaviSoni) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/Booking.hs:47]: there is one to many mapping for fromLocationId and toLocationId as these are same for searchReq , booking , ride so they will be there if created at the starting

**Review by hkmangla — APPROVED** [review]
